README.md
# Yapay SDK


[![Crate version on crates.io](https://img.shields.io/crates/v/yapay-sdk-rust)](https://crates.io/crates/yapay-sdk-rust)
[![Crate documentation on docs.rs](https://img.shields.io/docsrs/yapay-sdk-rust)](https://docs.rs/yapay-sdk-rust)
![Crate license](https://img.shields.io/crates/l/yapay-sdk-rust)


<!-- cargo-rdme start -->

An open source, strongly-typed SDK for the Yapay API.

It will try to hold your hand and reduce the possibility of errors, providing the correct API
surface.

### Note

The library is still under development and its public API is subject to change.

# Installation

Added the following into your Cargo.toml:

```toml
yapay_sdk_rust = "0.1"
```

# Usage

The client is built using the
[`YapaySDKBuilder::with_token`](https://docs.rs/yapay-sdk-rust/latest/yapay_sdk_rust/struct.YapaySDKBuilder.html) `with_token`
method.

```rust
use yapay_sdk_rust::{YapaySDK, YapaySDKBuilder};

let yapay_sdk: YapaySDK = YapaySDKBuilder::with_token(env!("YAPAY_ACCOUNT_TOKEN"));

```

Once the token is inserted, you can call methods on [`crate::YapaySDK`]



# Creating a Checkout link

You can easily retrieve a checkout link with the method below.

```rust
use std::num::NonZeroU8;

use uuid::Uuid;
use yapay_sdk_rust::{
    CheckoutPreferences, PaymentCreditCard, YapayEnv, YapayProduct, YapaySDKBuilder,
};

#[tokio::main]
async fn async_main() {
    // your token, can come from environment or else
    let yapay_token = "YAPAY_ACCOUNT_TOKEN";
    let yapay_sdk = YapaySDKBuilder::with_token(yapay_token);

    let product = YapayProduct::new(
        "note-100sk".to_string(),
        "Notebook Cinza".to_string(),
        NonZeroU8::new(1).unwrap(),
        2453.50,
    );

    let order_number = Uuid::new_v4().to_string();
    let checkout_preferences = CheckoutPreferences::new(order_number, vec![product])
        .expect("Validation failed.")
        .set_notification_url("https://your-notifications-url.com")
        .expect("Notifications URL failed to validate.")
        .set_available_payment_methods(&PaymentCreditCard::payment_methods_all());

    let checkout_url = yapay_sdk
        .create_checkout_page(YapayEnv::PRODUCTION, checkout_preferences)
        .await
        .expect("Something went wrong creating the checkout.");
}
```

# Other Examples

Check out the `tests` folder inside our repository to check for more examples.

# License
Project is licensed under the permissive MIT license.

<!-- cargo-rdme end -->


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## v0.3.3 (2022-05-31)

### Refactor

 - <csr-id-21d751ffdc75d1552a78d455ce81294fc2fc9811/> clippy lints, re-export of other missing Customer structs


### Commit Statistics

<csr-read-only-do-not-edit/>

 - 1 commit contributed to the release.
 - 1 commit where understood as [conventional](https://www.conventionalcommits.org).
 - 0 issues like '(#ID)' where seen in commit messages

### Commit Details

<csr-read-only-do-not-edit/>

<details><summary>view details</summary>

 * **Uncategorized**
    - clippy lints, re-export of other missing Customer structs ([`21d751f`](https://github.comgit//saskenuba/yapay-sdk-rust/commit/21d751ffdc75d1552a78d455ce81294fc2fc9811))
</details>

## v0.3.2 (2022-05-31)

<csr-id-a90f86f702a16981fb992e263b8acdd78becd235/>

### Refactor

 - <csr-id-a90f86f702a16981fb992e263b8acdd78becd235/> TransactionStatus -> YapayTransactionStatus, add re-export


### Commit Statistics

<csr-read-only-do-not-edit/>

 - 2 commits contributed to the release.
 - 1 day passed between releases.
 - 1 commit where understood as [conventional](https://www.conventionalcommits.org).
 - 0 issues like '(#ID)' where seen in commit messages

### Commit Details

<csr-read-only-do-not-edit/>

<details><summary>view details</summary>

 * **Uncategorized**
    - Release yapay-sdk-rust v0.3.2 ([`fa5fa6b`](https://github.comgit//saskenuba/yapay-sdk-rust/commit/fa5fa6b3db98bca78fda6bc2a62d8c4f47ff9df5))
    - TransactionStatus -> YapayTransactionStatus, add re-export ([`a90f86f`](https://github.comgit//saskenuba/yapay-sdk-rust/commit/a90f86f702a16981fb992e263b8acdd78becd235))
</details>

## v0.3.1 (2022-05-29)

<csr-id-bea195316e37402674108bf2c7b95101d7c07662/>

### Refactor

 - <csr-id-bea195316e37402674108bf2c7b95101d7c07662/> removed print calls that were panicking on multi-thread..
   executors, and replaced with tracing dep

### Commit Statistics

<csr-read-only-do-not-edit/>

 - 2 commits contributed to the release.
 - 1 commit where understood as [conventional](https://www.conventionalcommits.org).
 - 0 issues like '(#ID)' where seen in commit messages

### Commit Details

<csr-read-only-do-not-edit/>

<details><summary>view details</summary>

 * **Uncategorized**
    - Release yapay-sdk-rust v0.3.1 ([`a14714d`](https://github.comgit//saskenuba/yapay-sdk-rust/commit/a14714df9def5d8c1114cbe684303969cc45f137))
    - removed print calls that were panicking on multi-thread.. ([`bea1953`](https://github.comgit//saskenuba/yapay-sdk-rust/commit/bea195316e37402674108bf2c7b95101d7c07662))
</details>

## v0.3.0 (2022-05-28)

<csr-id-9e66ae0066fe1789d2f8b5a52a83996c0ccc4f4d/>
<csr-id-c2f8fa7d808a09dbcba4f6cfe9878759ed930ea9/>

### New Features

 - <csr-id-3daa7dffe991c1143b2e0b77e7bf49602438ab77/> implemented TryFrom<Vec<u8>> for `YapayWebhook` to allow it...
   to be extracted from raw payloads.

### Other

 - <csr-id-9e66ae0066fe1789d2f8b5a52a83996c0ccc4f4d/> removed unused deps and traits


### Chore (BREAKING)

 - <csr-id-c2f8fa7d808a09dbcba4f6cfe9878759ed930ea9/> removed pub visibilty to allow proper re-export of structs


### Commit Statistics

<csr-read-only-do-not-edit/>

 - 4 commits contributed to the release.
 - 3 commits where understood as [conventional](https://www.conventionalcommits.org).
 - 0 issues like '(#ID)' where seen in commit messages

### Commit Details

<csr-read-only-do-not-edit/>

<details><summary>view details</summary>

 * **Uncategorized**
    - Release yapay-sdk-rust v0.3.0 ([`5d75b68`](https://github.comgit//saskenuba/yapay-sdk-rust/commit/5d75b6810d1d0d6c9369179bfeb12acbd7017205))
    - removed pub visibilty to allow proper re-export of structs ([`c2f8fa7`](https://github.comgit//saskenuba/yapay-sdk-rust/commit/c2f8fa7d808a09dbcba4f6cfe9878759ed930ea9))
    - implemented TryFrom<Vec<u8>> for `YapayWebhook` to allow it... ([`3daa7df`](https://github.comgit//saskenuba/yapay-sdk-rust/commit/3daa7dffe991c1143b2e0b77e7bf49602438ab77))
    - removed unused deps and traits ([`9e66ae0`](https://github.comgit//saskenuba/yapay-sdk-rust/commit/9e66ae0066fe1789d2f8b5a52a83996c0ccc4f4d))
</details>

## v0.2.0 (2022-05-28)

### Documentation

 - <csr-id-6636633ec54386822205b3ce1135c4a1d14d5743/> updated README
 - <csr-id-b8bdaedef278163ffe4115d896d2c69e6f4b18c3/> updated docs/readme

### New Features

 - <csr-id-3d1ffe0e1f1e1cae75fe713a7909334f7bbe30fc/> Added `set_process_url` builder fn
 - <csr-id-abbb13b5acbe107767ca45b4abba3f2a7782e71c/> added TransactionStatus enum, and replaced it on webhook

### Bug Fixes

 - <csr-id-c0e57a7ea452bea6580e8e16169071c9504c8268/> clippy lints

### New Features (BREAKING)

 - <csr-id-08bb5f2bbbb497fc15b4a9ebdef69498a37db536/> docs, fixes +
   * Added tests
* Updated some functions docs;
* Checkout can handle available_payment_methods correctly now;
* Renamed Webhook to YapayWebhook;

### Commit Statistics

<csr-read-only-do-not-edit/>

 - 7 commits contributed to the release over the course of 3 calendar days.
 - 3 days passed between releases.
 - 6 commits where understood as [conventional](https://www.conventionalcommits.org).
 - 0 issues like '(#ID)' where seen in commit messages

### Commit Details

<csr-read-only-do-not-edit/>

<details><summary>view details</summary>

 * **Uncategorized**
    - Release yapay-sdk-rust v0.2.0 ([`353a644`](https://github.comgit//saskenuba/yapay-sdk-rust/commit/353a6440e47b857a7e60a2e178fa991f790423ad))
    - updated README ([`6636633`](https://github.comgit//saskenuba/yapay-sdk-rust/commit/6636633ec54386822205b3ce1135c4a1d14d5743))
    - Added `set_process_url` builder fn ([`3d1ffe0`](https://github.comgit//saskenuba/yapay-sdk-rust/commit/3d1ffe0e1f1e1cae75fe713a7909334f7bbe30fc))

src/lib.rs
//! An open source, strongly-typed SDK for the Yapay API.
//!
//! It will try to hold your hand and reduce the possibility of errors, providing the correct API
//! surface.
//!
//! ### Note
//!
//! The library is still under development and its public API is subject to change.
//!
//! # Installation
//!
//! Added the following into your Cargo.toml:
//!
//! ```toml
//! yapay_sdk_rust = "0.1"
//! ```
//!
//! # Usage
//!
//! The client is built using the
//! [`YapaySDKBuilder::with_token`](crate::YapaySDKBuilder) `with_token`
//! method.
//!
//! ```rust
//! # fn main() {
//! use yapay_sdk_rust::{YapaySDK, YapaySDKBuilder};
//!
//! let yapay_sdk: YapaySDK = YapaySDKBuilder::with_token(env!("YAPAY_ACCOUNT_TOKEN"));
//!
//! # }
//! ```
//!
//! Once the token is inserted, you can call methods on [`crate::YapaySDK`]
//!
//!
//!
//! # Creating a Checkout link
//!
//! You can easily retrieve a checkout link with the method below.
//!
//! ```no_run
//! use std::num::NonZeroU8;
//!
//! use uuid::Uuid;
//! use yapay_sdk_rust::{
//!     CheckoutPreferences, PaymentCreditCard, YapayEnv, YapayProduct, YapaySDKBuilder,
//! };
//!
//! #[tokio::main]
//! async fn async_main() {
//!     // your token, can come from environment or else
//!     let yapay_token = "YAPAY_ACCOUNT_TOKEN";
//!     let yapay_sdk = YapaySDKBuilder::with_token(yapay_token);
//!
//!     let product = YapayProduct::new(
//!         "note-100sk".to_string(),
//!         "Notebook Cinza".to_string(),
//!         NonZeroU8::new(1).unwrap(),
//!         2453.50,
//!     );
//!
//!     let order_number = Uuid::new_v4().to_string();
//!     let checkout_preferences = CheckoutPreferences::new(order_number, vec![product])
//!         .expect("Validation failed.")
//!         .set_notification_url("https://your-notifications-url.com")
//!         .expect("Notifications URL failed to validate.")
//!         .set_available_payment_methods(&PaymentCreditCard::payment_methods_all());
//!
//!     let checkout_url = yapay_sdk
//!         .create_checkout_page(YapayEnv::PRODUCTION, checkout_preferences)
//!         .await
//!         .expect("Something went wrong creating the checkout.");
//! }
//! ```
//!
//! # Other Examples
//!
//! Check out the `tests` folder inside our repository to check for more examples.
//!
//! # License
//! Project is licensed under the permissive MIT license.

#![allow(
    clippy::missing_const_for_fn,
    clippy::missing_errors_doc,
    clippy::missing_panics_doc,
    clippy::must_use_candidate,
    clippy::non_ascii_literal,
    clippy::redundant_closure,
    clippy::use_self,
    clippy::used_underscore_binding
)]
#![warn(missing_debug_implementations, missing_copy_implementations)]
#![deny(
    trivial_casts,
    trivial_numeric_casts,
    unused_import_braces,
    unused_qualifications
)]

mod checkout;
mod common_types;
pub mod errors;
mod helpers;
mod simulation;
mod transaction;
mod webhooks;

use std::marker::PhantomData;

pub use checkout::CheckoutPreferences;
use common_types::ResponseRoot;
pub use common_types::{
    AddressType, AsPaymentMethod, CustomerAddress, CustomerPhoneContact, PaymentCreditCard,
    PaymentOtherMethods, PhoneContactType, YapayCardData, YapayCustomer, YapayProduct,
    YapayTransaction, YapayTransactionStatus,
};
use futures::TryFutureExt;
use reqwest::header::{CONTENT_TYPE, LOCATION};
use reqwest::redirect::Policy;
use reqwest::{Client, Method};
use serde::de::DeserializeOwned;
use serde::Serialize;
use validator::Validate;
pub use webhooks::YapayWebhook;

use crate::errors::{ApiError, InvalidError, SDKError};
use crate::simulation::{PaymentTaxResponse, SimulatePayload, SimulationResponseWrapper};
use crate::transaction::creditcard::TransactionResponse;
use crate::transaction::{PaymentRequestRoot, TransactionResponseWrapper};

const API_PROD_BASE: &str = "https://api.intermediador.yapay.com.br/api";
const API_TEST_BASE: &str = "https://api.intermediador.sandbox.yapay.com.br/api";

const CHECKOUT_PROD_BASE: &str = "https://tc.intermediador.yapay.com.br/payment/transaction";
const CHECKOUT_TEST_BASE: &str =
    "https://tc-intermediador-sandbox.yapay.com.br/payment/transaction";

pub trait CanValidate: Serialize + Validate {}

#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub enum YapayEnv {
    PRODUCTION,
    SANDBOX,
}

impl YapayEnv {
    pub const fn checkout_link(self) -> &'static str {
        match self {
            Self::PRODUCTION => CHECKOUT_PROD_BASE,
            Self::SANDBOX => CHECKOUT_TEST_BASE,
        }
    }

    pub const fn api_link(self) -> &'static str {
        match self {
            Self::PRODUCTION => API_PROD_BASE,
            Self::SANDBOX => API_TEST_BASE,
        }
    }
}

///
#[derive(Copy, Clone, Debug)]
pub struct YapaySDKBuilder {}

impl YapaySDKBuilder {
    /// Creates an [`YapaySDK`] ready to request the API.
    pub fn with_token<T>(account_token: &T) -> YapaySDK
    where
        T: ToString,
    {
        let http_client = Client::builder()
            .cookie_store(true)
            .redirect(Policy::none())
            .build()
            .expect("Failed to create client.");

        YapaySDK {
            http_client,
            account_token: account_token.to_string(),
        }
    }
}

#[derive(Debug)]
pub struct YapaySDK {
    pub(crate) http_client: Client,
    pub(crate) account_token: String,
}

#[derive(Debug)]
pub struct SDKJsonRequest<'a, RP> {
    http_client: &'a Client,
    method: Method,
    endpoint: &'a str,
    payload: String,
    response_type: PhantomData<RP>,
}

impl<'a, RP> SDKJsonRequest<'a, RP> {
    #[must_use]
    pub fn from_sdk(sdk: &'a YapaySDK, method: Method, endpoint: &'a str, payload: String) -> Self {
        Self {
            http_client: &sdk.http_client,
            method,
            endpoint,
            response_type: Default::default(),
            payload,
        }
    }
}

impl<'a, RP> SDKJsonRequest<'a, RP> {
    /// Injects bearer token, and return response
    pub async fn execute(self, yapay_env: YapayEnv) -> Result<RP, SDKError>
    where
        RP: DeserializeOwned + Send,
    {
        let api_endpoint = format!("{}{}", yapay_env.api_link(), self.endpoint);
        tracing::trace!("api endpoint: {:?}", api_endpoint);

        let request = self
            .http_client
            .request(self.method, api_endpoint)
            .body(self.payload)
            .header(CONTENT_TYPE, "application/json")
            .build()
            .unwrap();
        tracing::trace!("request = {:#?}", request);

        let response = self
            .http_client
            .execute(request)
            .and_then(reqwest::Response::text)
            .await?;
        tracing::trace!("response = {}", response);

        // matches errors due to wrong payloads etc
        let error_jd = serde_json::from_str::<ApiError>(&*response);
        if let Ok(err) = error_jd {
            tracing::error!("err = {:#?}", err);
            return Err(SDKError::PayloadError(err));
        }

        let jd = &mut serde_json::Deserializer::from_str(&*response);
        let res: Result<RP, _> = serde_path_to_error::deserialize(jd);

        match res {
            Ok(deserialized_resp) => Ok(deserialized_resp),
            Err(err) => {
                tracing::error!("{:?}", err.path());
                tracing::error!("Error = {:#?}", err);
                Err(SDKError::GenericError)
            }
        }
    }
}

pub type CardTransactionResponse = ResponseRoot<TransactionResponseWrapper<TransactionResponse>>;
pub type SimulationResponse = ResponseRoot<SimulationResponseWrapper<PaymentTaxResponse>>;

impl YapaySDK {
    pub async fn create_checkout_page(
        &self,
        yapay_env: YapayEnv,
        checkout_preferences: CheckoutPreferences,
    ) -> Result<String, SDKError> {
        let querystring = checkout_preferences.to_form(&*self.account_token);
        let request = self
            .http_client
            .request(Method::POST, yapay_env.checkout_link())
            .header(CONTENT_TYPE, "application/x-www-form-urlencoded")
            .body(querystring)
            .build()
            .unwrap();

        let response = self.http_client.execute(request).await.unwrap();

        response
            .headers()
            .get(LOCATION)
            .and_then(|hdr| hdr.to_str().ok())
            .map(ToString::to_string)
            .ok_or(SDKError::GenericError)
    }

    /// Returns an error if it fails to validate any of its arguments.
    pub fn create_credit_card_payment(
        &self,
        customer: YapayCustomer,
        transaction: YapayTransaction,
        products: Vec<YapayProduct>,
        cc_payment_data: YapayCardData,
    ) -> Result<SDKJsonRequest<CardTransactionResponse>, SDKError> {
        let request_payload = PaymentRequestRoot::new(
            self.account_token.clone(),
            customer,
            products,
            transaction,
            cc_payment_data,
        );

        if let Err(errs) = request_payload.validate() {
            return Err(InvalidError::ValidatorLibError(errs).into());
        }

        let payload = serde_json::to_string(&request_payload).expect("Safe to unwrap.");

        Ok(SDKJsonRequest::from_sdk(
            self,
            Method::POST,
            "/v3/transactions/payment",
            payload,
        ))
    }

    #[must_use]
    pub fn simulate_payment(&self, total_amount: f64) -> SDKJsonRequest<SimulationResponse> {
        let request_payload = SimulatePayload::new(self.account_token.clone(), total_amount);
        let payload = serde_json::to_string(&request_payload).unwrap();

        SDKJsonRequest::from_sdk(
            self,
            Method::POST,
            "/v1/transactions/simulate_splitting",
            payload,
        )
    }
}
